openpype/plugins/publish/integrate_new.py

```diff
@@ -31,6 +27,175 @@
+                # Backup original file
+                # todo: add timestamp or uuid to ensure unique
+                backup = dst + ".bak"
+                self._backup_to_original[backup] = dst
+                self.log.debug("Backup existing file: "
+                               "{} -> {}".format(dst, backup))
+                os.rename(dst, backup)
+
+        # Copy the files to transfer
+        for dst, (src, opts) in self._transfers.items():
+            self._create_folder_for_file(dst)
+
+            if opts["mode"] == self.MODE_COPY:
+                self.log.debug("Copying file ... {} -> {}".format(src, dst))
+                copyfile(src, dst)
+            elif opts["mode"] == self.MODE_HARDLINK:
+                self.log.debug("Hardlinking file ... {} -> {}".format(src,
+                                                                      dst))
+                create_hard_link(src, dst)
+
+            self._transferred.append(dst)
+
+    def finalize(self):
+        # Delete any backed up files
+        for backup in self._backup_to_original.keys():
+            try:
+                os.remove(backup)
+            except OSError:
+                self.log.error("Failed to remove backup file: "
+                               "{}".format(backup),
+                               exc_info=True)
+
+    def rollback(self):
+
+        errors = 0
+
+        # Rollback any transferred files
+        for path in self._transferred:
+            try:
+                os.remove(path)
+            except OSError:
+                errors += 1
+                self.log.error("Failed to rollback created file: "
+                               "{}".format(path),
+                               exc_info=True)
+
+        # Rollback the backups
+        for backup, original in self._backup_to_original.items():
+            try:
+                os.rename(backup, original)
+            except OSError:
+                errors += 1
+                self.log.error("Failed to restore original file: "
+                               "{} -> {}".format(backup, original),
+                               exc_info=True)
+
+        if errors:
+            self.log.error("{} errors occurred during "
+                           "rollback.".format(errors), exc_info=True)
+            six.reraise(*sys.exc_info())
+
+    @property
+    def transferred(self):
+        """Return the processed transfers destination paths"""
+        return list(self._transferred)
+
+    @property
+    def backups(self):
+        """Return the backup file paths"""
+        return list(self._backup_to_original.keys())
+
+    def _create_folder_for_file(self, path):
+        dirname = os.path.dirname(path)
+        try:
+            os.makedirs(dirname)
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                pass
+            else:
+                self.log.critical("An unexpected error occurred.")
+                six.reraise(*sys.exc_info())
```

Comment:
I would simplify this, all the try->except does is reraise exception.
```suggestion
        if not os.path.exists(dirname):
            os.makedirs(dirname)
```

Reply:
I believe the original behavior came from the fact that _technically_ the folder could have been created in-between the call to `os.makedirs` and `os.path.exists` and thus the `try..except` is the only way to correctly capture the folder didn't exist at the time of the call to `os.makedirs`. I know it's an extreme edge case, but it's considered safer as it currently is.

If we were python 3.2+ (which got further fixes in 3.4.1+) everywhere we could rely on `os.makedirs(dirname, exist_ok=True)` but unfortunately we aren't.